lib/inv_bicgstab_quda.cpp

```diff
@@ -50,34 +41,40 @@ namespace quda {
 
   void BiCGstab::operator()(ColorSpinorField &x, ColorSpinorField &b)
   {
-    profile.TPSTART(QUDA_PROFILE_PREAMBLE);
+    if (!param.is_preconditioner) profile.TPSTART(QUDA_PROFILE_INIT);
+
+    double b2 = blas::norm2(b); // norm sq of source
+    double r2 = 0.0;            // norm sq of residual
+
+    // Check to see that we're not trying to invert on a zero-field source
+    if (b2 == 0) {
+      if (param.compute_null_vector == QUDA_COMPUTE_NULL_VECTOR_NO) {
+        warningQuda("inverting on zero-field source");
+        x = b;
+        param.true_res = 0.0;
+        param.true_res_hq = 0.0;
+        profile.TPSTOP(QUDA_PROFILE_INIT);
+        return;
+      } else if (param.use_init_guess == QUDA_USE_INIT_GUESS_YES) {
+        b2 = r2;
+      } else {
+        errorQuda("Null vector computing requires non-zero guess!");
+      }
+    }
 
     if (!init) {
```

Comment:
Can you align this cleanup of BiCGStab with the cleanup in other solvers already done: namely field allocation should be done in the virtual `init` function?  Also add the `get_residual()` method to allow use as a pre-smoother.

Reply:
https://github.com/lattice/quda/pull/1421/commits/53dced96630d2269daa8f03eef60ff83c7a8dcac

Reply:
https://github.com/lattice/quda/pull/1421/commits/53dced96630d2269daa8f03eef60ff83c7a8dcac